### DISCUSSION TITLE
请教一下一些 DBus 接口的含义

### DISCUSSION THREAD

**Discussion by tonyfettes** [fcitx/fcitx5#397] [Q&A] [answered]
我在使用 DBus 接口的过程中遇到了一些不熟悉的概念。因此想请教下列问题：

1. 什么是 Auxiliary Up/Down？他们有什么用？
我在 fcitx 维基上找到了[这个页面](https://fcitx-im.org/wiki/Create_a_new_Fcitx_Theme)，其中描述了 Auxiliary Up/Down 的位置。但是我仍然不知道这两个有什么作用。

2. DBus 接口中 SetCursorRect/SetCursorRectV2 这两个方法的含义
我猜测这两个方法是用来通知 Fcitx5 图形界面中鼠标指针位置的。https://github.com/fcitx/fcitx5-gtk/blob/master/fcitx-gclient/fcitxgclient.c#L447 里面的注释说 `x,y,w,h` 分别指鼠标指针的横纵座标和宽度高度，没有给窗口信息的接口。所以这两个方法只是为了在 Fcitx5 的窗口里面绘制光标吗？

3. 我应该怎么选择传给 Fcitx5 的 SurroundingText？
a. 我应该传多少 SurroundingText ？
b. 光标前后的文字都需要吗？
c. 每次文字变更都需要调用一遍吗？

谢谢！

**Comment by wengxt** [comment, marked as answer]
1. 就是弹出的输入框中的一段文字，那个页面描述的是典型的摆放位置
2. 不是鼠标指针，是输入光标，就是你正在输入的位置，目的是让 fcitx 能把输入窗口移动到正确的位置
3. a. 传多少随意，但一般来说我在实现 im module 的时候都会限制大小，没有输入法会用到非常多
    b. 需要
    c. 是的
你可以选择不传，只需要不设置 capabilityflag 的对应位即可。

**Reply by tonyfettes** [reply]
明白了。谢谢您的回答。

**Reply by wengxt** [reply]
另外看你想做的事情，似乎是需要 client side input panel，如果这样的话，SetCursorRect 就不相关了，就不用琢磨那两个方法。

**Reply by tonyfettes** [reply]
行，谢谢您。
